Tags autocomplete: fix stale results appearing in large libraries, add pref for fuzzy search

abaevbog opened the pull request:
- do not add autocomplete results for queries that are no longer relevant. For example, if the user
types "test", and row handler is still called with results for "te" - ignore them. Otherwise, results with "te" but not "test" may appear in libraries with many tags
- add optional fuzzy search for tags with a visible pref checkbox

Fixes: #1647 

---

To look into how fuzzy tags autocomplete would perform now, I created a collection with 10,000 items, each with 100 random tags (example: `8-6QqFw-ULQ44-TG9KD`). I assume a library with 1,000,000 tags total should be close to the upper bound of the number of tags one would have because then I also started running into other performance issues, like a longer delay when switching between collections. For reference, this is the [exported RDF file of that collection](https://www.dropbox.com/scl/fi/6ddrfav0lunglsvwxukxc/million_tags_collection.rdf?rlkey=db7tryqbc330xanftv4lvvpff&st=r6358o5c&dl=0).

With so many tags, on current `main`, the autocomplete runs fast but it also includes stale search results that are irrelevant for the current query:

https://github.com/user-attachments/assets/3e9c14af-b6d5-4343-8f19-042d5c745075

When we only include results fetched for the latest input (as in this PR), I get search results after ~2 seconds if fuzzy matching is disabled:

https://github.com/user-attachments/assets/f38f3c29-5e74-4350-a5fe-5f6fc2882cd8


With fuzzy search results, it takes a bit longer but not by that much - 3 seconds or so:

https://github.com/user-attachments/assets/e0e3d528-b78e-4dd3-99da-ffb7c52d360d


We could consider more optimization with SQLite FTS as described in the linked issue - let me know if that would be worth it!

dstillman commented on chrome/content/zotero/zotero-autocomplete.js:
```diff
@@ -246,12 +247,19 @@ ZoteroAutoComplete.prototype.startSearch = Zotero.Promise.coroutine(function* (s
+			if (currentSearchString != this._result.searchString) {
+				cancel();
+				return;
+			}
```
This is what `stopSearch()` — currently unused with a couple `FIXME`s — is supposed to be doing, so we should clean that up at the same time if this logic is just better. And to clarify, this isn't just "ignoring" results from other queries — it's actually stopping the SQLite query (but see my other note about this).

dstillman commented on defaults/preferences/zotero.js:
```diff
@@ -28,6 +28,7 @@ pref("extensions.zotero.downloadAssociatedFiles",true);
 pref("extensions.zotero.findPDFs.resolvers", '[]');
 pref("extensions.zotero.reportTranslationFailure",true);
 pref("extensions.zotero.automaticTags",true);
+pref("extensions.zotero.tagsFuzzyAutocomplete", false);
```
Let's just optimize this as best as we can. No need to pref it (unless we start getting reports that it's causing problems).

dstillman commented on chrome/locale/en-US/zotero/preferences.ftl:
```diff
@@ -59,6 +59,9 @@ preferences-locate-library-lookup-intro = Library Lookup can find a resource onl
 preferences-locate-base-url = Base URL:
 
+preferences-tags-fuzzy-autocomplete =
+    .label = Fuzzy tags autocomplete
```
(This wouldn't be "fuzzy" matching. We could actually do that, but that would imply finding strings that were similar but not exact. This is just a non-left-bound search.)

dstillman reviewed (commented):
One optimization we could do here would be to store matching tags in memory after the first query and then simply filter that list in JS, as long as the search string contains the original string (and has the same libraryID/itemID params). We'd then store the next list, and so on. We'd set a short timer to clean up old lists.

The one catch is that (despite my other comment) we wouldn't actually want to cancel the initial query anymore — we'd want to let it run and avoid running additional queries for supersets of the running query string, so that we could filter the results as soon as they were available.

To the user, this would hopefully just mean that the autocomplete list might not appear after the first few characters but then would start appearing later as it was filtering successively smaller lists.

abaevbog committed:
tag autocomplete matching non-left bound substring

Tag autocomplete will now match tag substrings in the middle,
not just the start of the tag.

Refactoring to optimize performance for libraries
with many tags:
- From each search, save results into an in-memory cache.
During subsequent searches, check the cache to see if
it has any matches
- Ensure we never run multiple queries at the
same time. If a query is already running, wait for it to
finish before running the next one. If there are multiple
pending queries that arrive while one is running,
discard all but the last one
- Fixed stopSearch not working correctly. Now, if
stopSearch is called, the current query processing will be
properly cancelled in the onRow handler.

Also, drop resultsCallback to keep function flow clearer, since it
is not used anywhere at the moment.

abaevbog commented:
I think something like this would work here! These are the optimizations I added:

- there is an in-memory cache that gets cleared out after a while. Before running a new query, we check that cache to see if there is anything matching the new query. It is stored as an object `{"searchString": ["matchOne", "matchTwo"]}`. So if we are searching for `"test"`, we'll check cached results for `"tes"`, `"te"`, and `"t"` if they exist. Those will be found fast and immediately added into the popup
- we still have to run the SQL query after. As described above, I attempted to load the entire superset of a shorter query into cache (by removing the `LIMIT` from SQL) - but that can take too long for small queries. Without `LIMIT`, queries for one or two characters take forever. Because of this, cache may not always have all the matches but it serves mainly to provide a faster visual feedback.
- after the SQL runs, any additional results are added into the popup and stored in cache as well for subsequent queries

One issue I noticed is that we ended up spawning multiple SQL queries simultaneously, since `stopSearch` didn't really work (e.g. "t", "te", "tes", "test" would all start a new query, regardless if the old one finished). It ended up affecting performance after typing a few characters, and it also meant that the autocomplete searches kept running for a while even after the autocomplete popup was closed.

To tackle that, I added proper `stopSearch` handling and separated `startSearch` (called when user types) from `executeQuery` that runs the actual search to ensure that we only run one query at a time. If there are multiple queries that are send by the user while a query is already loading, the last query will be executed after the first one is finished (via cancellation or otherwise), and all other ones will be discarded.

This is an example of events now:
- user types "t" (`startSearch` called)
- SQL is being executed for searchString "t"
- user types "e"
- `stopSearch` may be called (but not always?), which will mark "t" query as cancelled. `startSearch` with "te" is called.
- user types "s" (`stopSearch` may be called, `startSearch` called)
- user types "t" (`stopSearch` may be called, `startSearch` called)
- SQL for "e" is finished via `cancel` in `onRow`
-  if there were any search results for "t" added into the cache that match "test", they will appear in the popup
- SQL is being executed for searchString "test" ("te", and "tes" are ignored)
- SQL finishes for "test", popup updated

This is what the search looks like on my end with 1,000,000 tags library:

https://github.com/user-attachments/assets/2aa88f79-f624-4d38-87f1-ba2f5e1bbb29
